File: 2347-count-nodes-equal-to-average-of-subtree/count-nodes-equal-to-average-of-subtree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfSubtree(self, root: Optional[TreeNode]) -> int:

        ans =0
      
        def rec(root):
            if root ==None:
                return (0,0)
            nonlocal ans 
            l_sum , l_count = rec(root.left)
            r_sum,r_count = rec(root.right)
            if math.floor((l_sum+r_sum+root.val) / (l_count + r_count+1)) == root.val:
                ans+=1
            return (l_sum+r_sum+root.val,l_count + r_count+1 )
        rec(root)
        return ans



        

File: 2347-count-nodes-equal-to-average-of-subtree/count-nodes-equal-to-average-of-subtree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans;
    private int[] rec(TreeNode root){

        if (root  == null){

            return new int[]{0,0};

        }

        int [] leftResult= rec(root.left);
        int [] rightResult = rec(root.right);

        int sum  =leftResult[0]+rightResult[0]+root.val ; 
        int count =  leftResult[1]+rightResult[1]+1;

        if( Math.floorDiv(sum,count) == root.val)
            ans++;

        return new int[]{sum, count};

    }
    public int averageOfSubtree(TreeNode root) {
        rec(root);
        return ans;
        
    }
}